sdk/scheduler/src/main/java/com/mesosphere/sdk/http/types/PlanInfo.java

```diff
@@ -23,11 +22,16 @@
     private final Status status;
 
     public static PlanInfo forPlan(final Plan plan) {
-        List<PhaseInfo> phaseInfos = plan.getChildren().stream()
+        // Calculate the status of the plan based on the phases, THEN generate the PhaseInfos for those phases.
+        // This ordering is a workaround for potential inconsistency when step statuses change while we're rendering the
+        // plan. By fetching phase statuses after the plan status, inconsistencies should typically appear as e.g. a
+        // plan that's IN_PROGRESS when the steps or phases are COMPLETE. If we did the plan status last, then we'd risk
+        // getting the opposite of that, which is less intuitive to an end user.
+        Status planStatus = plan.getStatus(); // plan status first ...
+        List<PhaseInfo> phaseInfos = plan.getChildren().stream() // ... then phases
                 .map(phase -> PhaseInfo.forPhase(phase))
                 .collect(Collectors.toList());
-
-        return new PlanInfo(phaseInfos, plan.getStrategy().getName(), plan.getErrors(), plan.getStatus());
+        return new PlanInfo(phaseInfos, plan.getStrategy().getName(), plan.getErrors(), planStatus);
```

Comment:
`plan.getErrors()` also visits the children. Could a similar ordering issue not occur when displaying an `ERROR` vs `COMPLETED` state?

Reply:
It is very unlikely (impossible in today's world) that the errors will ever change, but it's in principle the same kind of mistake we're trying to fix here.  A plan's status and errors should probably be grabbed as close together as possible.  There's still a race there.  Grab the errors first, then the status could produce non-empty error list and _non-ERROR_ status.   Grab the status first and get an ERROR status, but then grab errors and get an _empty_ ERROR list is also possible.  I think there's probably a few other bad combinations.  We're punting on correct consistency anyway, so I'm not too troubled either way.  None of the scenarios here seem more or less intuitive to me.